File: src/levels/L11-state/strategy.rb
# Una estrategia por evento… pero ninguna recuerda en qué estado está el pedido.
class Order
  attr_reader :status

  HANDLERS = { pagar: PayHandler, preparar: PrepareHandler, entregar: DeliverHandler, cancelar: CancelHandler }

  # region: EventHandlers
  def handle(event)
    @status = HANDLERS.fetch(event).new.apply # siempre aplica, pase lo que pase
    @history.record(self, event)
  end
  # endregion
end

File: src/levels/L11-state/ts/strategy.ts
// Una estrategia por evento… pero ninguna recuerda en qué estado está el pedido.
class Order {
  status = 'pendiente'
  private history = new History()

  static HANDLERS: Record<string, new () => { apply(): string }> = {
    pagar: PayHandler, preparar: PrepareHandler, entregar: DeliverHandler, cancelar: CancelHandler,
  }

  // region: EventHandlers
  handle(event: string) {
    this.status = new Order.HANDLERS[event]().apply() // siempre aplica, pase lo que pase
    this.history.record(this, event)
  }
  // endregion
}
